### PR TITLE
Rewrite8 changelog

### PR DESCRIPTION
**Pull request by mike-solomon**
There's a few issues I know about right now:

* The steps to convert a recipe to a scanning recipe are unclear. The example test we have to show the [conversion links to a non-existent guide](https://github.com/openrewrite/rewrite/blob/main/rewrite-java-test/src/test/java/org/openrewrite/java/recipes/MigrateRecipeToRewrite8Test.java#L542).
* The steps to convert JavaTemplate to the new one are unclear. In our example test, we say that [things "could be unnecessary" or that people should double-check this](https://github.com/openrewrite/rewrite/blob/main/rewrite-java-test/src/test/java/org/openrewrite/java/recipes/MigrateJavaTemplateToRewrite8Test.java#L121-L122) - but I am not sure how anyone would know the "right" way here. (**EDIT**: I attempted to make this clearer in the text)
* The steps to replace doNext similarly link to a[ non-existent guide](https://github.com/openrewrite/rewrite/blob/main/rewrite-java-test/src/test/java/org/openrewrite/java/recipes/MigrateRecipeToRewrite8Test.java#L485) and it's unclear how someone would know what to do.

It's also possible that other things I've said in the doc are misleading or wrong - so good to double check that.

If you want to see what this looks like in GitBook right now you can check out my test here: https://app.gitbook.com/o/-MEp_3EtccewzekKY8mZ/s/JC9dRbwVINQjAyoDyBuW/mike-test

### CODE REVIEW

**Comment by mike-solomon** [comment]
I'll add the generated recipes later on - didn't want to clutter this PR with those right now.

File: changelog/8-1-1-Release.md [hunk 1]
```diff
@@ -0,0 +1,1048 @@
...
+
+TODO: This migration recipe does not provide enough context. It links to a non-existant guide: https://github.com/openrewrite/rewrite/blob/main/rewrite-java-test/src/test/java/org/openrewrite/java/recipes/MigrateRecipeToRewrite8Test.java#L485
+
+## Migrating your recipes
+
+We've created a [migration recipe](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/recipes/MigrateRecipeToRewrite8.java) that will assist you with migrating your recipes to the latest version. You should run this recipe against all of your existing recipes. If the migration recipe is not able to fully migrate your recipe, comments will be added to the code to request a human to review it.
+
+Please feel free to reach out in the [community Slack](https://join.slack.com/t/rewriteoss/shared_invite/zt-nj42n3ea-b~62rIHzb3Vo0E1APKCXEA) if you have any questions.
+
+## New recipes
```
Comment (mike-solomon) [inline changelog/8-1-1-Release.md:462]: You can ignore everything from here down -- this is auto-generated from the markdown generator.

File: changelog/8-1-2-Release.md [hunk 2]
```diff
@@ -0,0 +1,1109 @@
...
+
+Scanning recipes have three phases:
+
+1. A scanning phase that collects information while making no new code changes. In this phase, the `scanner` is called for each source file and information is added to the `accumulator` that the recipe will need for future steps.
+    * For example, a recipe might want to detect whether a project is a Maven project or not. The `scanner` could detect a `pom.xml` file and add a flag to the `accumulator` so that future steps know this.
+2. An _optional_ generating phase where new files are created (if any are needed). In this phase, the `accumulator` can be accessed to determine whether or not a file should be created.
+3. An editing phase where the recipe makes changes (as it would before). Like the generating phase, an `accumulator` can be accessed to make changes – but you **can not** randomly access other source code or files.
+
+TODO: We need an example for this. Right now, the migration recipe links to nothing: https://github.com/openrewrite/rewrite/blob/main/rewrite-java-test/src/test/java/org/openrewrite/java/recipes/MigrateRecipeToRewrite8Test.java#L542
```
Comment (mike-solomon) [inline changelog/8-1-2-Release.md:141]: To help make reviewing easier - pointing out the TODOs that I feel like still need work.

File: changelog/8-1-2-Release.md [hunk 3]
```diff
@@ -0,0 +1,1109 @@
...
+}
+```
+{% endtab %}
+{% endtabs %}
+
+### The doNext method has been removed
+
+The [org.openrewrite.Recipe doNext(..)](https://github.com/openrewrite/rewrite/blob/v7.40.8/rewrite-core/src/main/java/org/openrewrite/Recipe.java#L306-L313) method has been removed. If a recipe won't make any changes, this should be replaced with a [ScanningRecipe](#new-approach-to-recipe-visiting). If it does make changes, it should use [TreeVisitor#doAfterVisit()](https://github.com/openrewrite/rewrite/blob/v7.40.8/rewrite-core/src/main/java/org/openrewrite/TreeVisitor.java#L146).
+
+TODO: This migration recipe does not provide enough context. It links to a non-existant guide: https://github.com/openrewrite/rewrite/blob/main/rewrite-java-test/src/test/java/org/openrewrite/java/recipes/MigrateRecipeToRewrite8Test.java#L485
```
Comment (mike-solomon) [inline changelog/8-1-2-Release.md:458]: To help make reviewing easier - pointing out the TODOs that I feel like still need work.

File: changelog/8-1-2-Release.md (hunk 2 above)
Comment (mike-solomon) [reply changelog/8-1-2-Release.md:141]: Thanks - I'll update this to include that example.

File: changelog/8-1-2-Release.md (hunk 3 above)
Comment (mike-solomon) [reply changelog/8-1-2-Release.md:458]: Should I call out that `doAfterVisit(Recipe)` has been removed then in combination with this? 

Also - I'm not sure I follow what you're saying. Were people using `doNext(Recipe)` to run the recipe they're currently in? Why would that be the case?

**Comment by mike-solomon** [comment]
@jsoref I updated the `expect.txt` to include the words it thinks are typos but I'm still getting an error here. Did I make a mistake with how I went about doing that?

**Comment by jsoref** [comment]
To answer @mike-solomon 's question...

> `expect.txt` generally only wants one case variant of a word and will suggest removing the current one if it needs a change.

> The idea is that if you say `MICROSOFT` is ok, you aren't really saying that `Microsoft` is ok, but if you say `Microsoft` is ok, you're probably also saying `MICROSOFT` is ok while not saying that `microsoft` is ok, but when you say `microsoft` is ok you're saying that all of them are ok.

When mike added `controlflow` but didn't remove `Controlflow`, he confused check-spelling. I need to fix it so that it honors the most relaxed form (here: `controlflow`) and merely warns about the other form(s) (here: `Controlflow`). I'll try to work on that soon.

File: changelog/8-1-2-Release.md [hunk 4]
```diff
@@ -0,0 +1,1109 @@
...
+
+{% hint style="warning" %}
+There is, unfortunately, no way for a YAML recipe to use `Preconditions`. We hope to support such a feature for them eventually, though.
+{% endhint %}
+
+**Example**:
+
+{% tabs %}
+{% tab title="Before" %}
+```java
```
[the remark this answers is not in the document]
Comment (kunli2) [reply changelog/8-1-2-Release.md:25]: don't worry about this for now. 
I mean if you run the doc generation task, for the recipe `org.openrewrite.java.recipes.MigrateRecipeToRewrite8`, there will be a md file generated at `build/docs/reference/recipes/java/recipes/migraterecipetorewrite8.md`, and in this file, there will be a example session with a diff, here is what I got it locally 

```diff
--- org/openrewrite/staticanalysis/RenamePrivateFieldsToCamelCase.java
+++ org/openrewrite/staticanalysis/RenamePrivateFieldsToCamelCase.java
@@ -4,0 +4,1 @@

import org.openrewrite.*;
+import org.openrewrite.internal.lang.Nullable;
import org.openrewrite.java.JavaIsoVisitor;
@@ -28,0 +29,1 @@

    private static class RenameNonCompliantNames extends JavaIsoVisitor<ExecutionContext> {
+
        @Override
@@ -29,3 +31,5 @@
    private static class RenameNonCompliantNames extends JavaIsoVisitor<ExecutionContext> {
        @Override
-       public JavaSourceFile visitJavaSourceFile(JavaSourceFile cu, ExecutionContext ctx) {
-           Map<J.VariableDeclarations.NamedVariable, String> renameVariablesMap = new LinkedHashMap<>();
-           Set<String> hasNameSet = new HashSet<>();
+       public  @Nullable J visit(@Nullable Tree tree, ExecutionContext ctx) {
+           if (tree instanceof JavaSourceFile) {
+               JavaSourceFile cu = (JavaSourceFile) tree;
+               Map<J.VariableDeclarations.NamedVariable, String> renameVariablesMap = new LinkedHashMap<>();
+               Set<String> hasNameSet = new HashSet<>();

@@ -33,3 +37,3 @@
            Set<String> hasNameSet = new HashSet<>();

-           getCursor().putMessage("RENAME_VARIABLES_KEY", renameVariablesMap);
-           getCursor().putMessage("HAS_NAME_KEY", hasNameSet);
-           super.visitJavaSourceFile(cu, ctx);
+               getCursor().putMessage("RENAME_VARIABLES_KEY", renameVariablesMap);
+               getCursor().putMessage("HAS_NAME_KEY", hasNameSet);
+               super.visit(cu, ctx);

@@ -37,7 +41,8 @@
            super.visitJavaSourceFile(cu, ctx);

-           renameVariablesMap.forEach((key, value) -> {
-               if (!hasNameSet.contains(value) && !hasNameSet.contains(key.getSimpleName())) {
-                   doAfterVisit(new RenameVariable<>(key, value));
-                   hasNameSet.add(value);
-               }
-           });
-           return cu;
+               renameVariablesMap.forEach((key, value) -> {
+                   if (!hasNameSet.contains(value) && !hasNameSet.contains(key.getSimpleName())) {
+                       doAfterVisit(new RenameVariable<>(key, value));
+                       hasNameSet.add(value);
+                   }
+               });
+           }
+           return super.visit(tree, ctx);
        }
```

Adding a diff here can make it clear to the readers. or we can just add the link to the `build/docs/reference/recipes/java/recipes/migraterecipetorewrite8.md` here is also fine.

File: changelog/8-1-2-Release.md (hunk 3 above)
Comment (kunli2) [reply changelog/8-1-2-Release.md:664]: > Should I call out that doAfterVisit(Recipe) has been removed then in combination with this?

Yes, I think so.

> Also - I'm not sure I follow what you're saying. Were people using doNext(Recipe) to run the recipe they're currently in? Why would that be the case?

Not to run the current recipe in `doNext(Recipe)`, but a different recipe

**Review by kunli2 — APPROVED** [review]

**Review by knutwannheden — APPROVED** [review]

File: SUMMARY.md [hunk 5]
```diff
@@ -76,6 +85,18 @@
     * [Search](reference/recipes/concourse/search/README.md)
       * [Find pinned resources by type](reference/recipes/concourse/search/findpinnedresource.md)
       * [Find privileged resource_type definitions.](reference/recipes/concourse/search/findprivilegedresourcetype.md)
+  * [Cucumber](reference/recipes/cucumber/README.md)
+    * [Jvm](reference/recipes/cucumber/jvm/README.md)
+      * [Cucumber to JUnit Test Suites](reference/recipes/cucumber/jvm/cucumbertojunitplatformsuite.md)
+      * [Cucumber-Java8 migration to Cucumber-Java](reference/recipes/cucumber/jvm/cucumberjava8tojava.md)
+      * [Drop SummaryPrinter](reference/recipes/cucumber/jvm/dropsummaryprinter.md)
+      * [Replace @Cucumber with @Suite](reference/recipes/cucumber/jvm/cucumberannotationtosuite.md)
+      * [Replace Cucumber-Java step definition regexes with Cucumber expressions](reference/recipes/cucumber/jvm/regextocucumberexpression.md)
```
[the remark this answers is not in the document]
Comment (mike-solomon) [reply SUMMARY.md:94]: added to list of allowed words

**Pull request merged**